manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+recursive call on "def" will consume stack space. However, expression
+trees tend to unbalanced on the right (lots of "Let" in sequence,
+rather than nested inside each other), so putting the call on "body"
+in tail position is an interesting improvement over the naive
+definition: it gives bounded stack space consumption if we assume
+a bound on the nesting depth of "Let" constructs.
+
+One would also get an error when using conflicting annotations, asking
+for two of the constructor arguments to be put in tail position:
+\begin{verbatim}[error]
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, (map_vars[@tailcall]) f def), (map_vars[@tailcall]) f body)
+
+Error: [@tail_mod_cons]: this constructor application may be TMC-transformed
+       in several different ways. Only one of the arguments may become a TMC
+       call, but several arguments contain calls that are explicitly marked
+       as tail-recursive. Please fix the conflict by reviewing and fixing the
+       conflicting annotations.
+Line 5, characters 15-42:   This call is explicitly annotated.
+Line 5, characters 45-73:   This call is explicitly annotated.
+\end{verbatim}
+
+\subsection{ss:out-of-tmc}{Danger: getting out of tail-mod-cons}
+
+Due to the nature of the tail-mod-cons transformation
+(see Section~\ref{ss:details} for a presentation of transformation):
+\begin{itemize}
+\item Calls from a tail-mod-cons function to another tail-mod-cons
+  function declared in the same recursive-binding group are ``tail
+  calls'', whether they occur in tail position or in tail-modulo-cons
+  positions.
+\item Calls from a function \emph{not} marked tail-mod-cons to
+  a tail-mod-cons function or, conversely, from a tail-mod-cons
+  function to a not-tail-mod-cons function are \emph{not} tail calls,
+  even if they syntactically appear in tail position in the source
+  program.
+\end{itemize}
+
+The fact that calls in tail position in the source program may become
+non-tail calls if they go from a tail-mod-cons to a non-tail-mod-cons
+function is surprising, and the transformation will warn about them.
+
+For example:
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+    let rec append_flatten xs xss =
+      match xs with
+      | [] -> flatten xss
+      | x :: xs -> x :: append_flatten xs xss
+    in append_flatten xs xss
+
+
+|                       append_flatten xs xss
+                        ^^^^^^^^^^^^^^^^^^^^^
+Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+but the function called is not itself specialized for TMC,
+so the call will not be transformed into a tail call.
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+Here the "append_flatten" helper is not annotated with
```

Comment:
```suggestion
Here the "append_flatten" helper is not marked with
```

Reply:
The chapter is using both "marked" and "annotated", I guess there was a subtle reason why one or the other felt more natural at the time of writing, but maybe we should use only one form? If so, I think we should go for the "annotated"; it is heavier but some use sites don't feel right with "marked" instead:

> [...] different from the evaluation order of the un-annotated version [...]

> Our recommendation is to use the "[\@tailcall]" annotation on
all callsites that should not consume any stack
space. "[\@tail_mod_cons]" extends the set of functions on which calls
can be annotated to be tail calls [...]

Reply:
Sounds reasonable to me - I think all I was after was the use of a single term (if that was then natural to use everywhere)